상점 API 개편

damiano1027 opened the pull request:
주요 개편 목록
- 상점 관련 DB 테이블 추가 및 정규화 flyway sql 파일 추가
- 개편된 DB 구조에 맞게 어드민 페이지 API 수정
- 일반 상점 API 및 사장님 전용 상점 API는 추후 완료 예정

damiano1027 committed:
feat: 상점 메뉴 마이그레이션 API

shop_menus 테이블 price_type 컬럼의 JSON 데이터들을 파싱해서 shop_menu_details 테이블로 이동시키는 API를 개발하였습니다.
- GET /admin/v2/shops/menus/migrate (koreatech.in.controller.v2.admin.AdminShopController)

damiano1027 committed:
feat: 어드민용 상점 메뉴 카테고리 CRUD API

사장님 상점 메뉴 CRUD하는 API를 개발하려면 메뉴 카테고리 API가 필요하여 개발하였습니다. admin 계정이면서 grant_shop 권한이 있어야 합니다.

damiano1027 committed:
feat: 사장님 상점 메뉴 Create, Read API

- 아직 request dto에 대한 데이터 유효성 검증은 하지 않습니다. 추후 추가 예정입니다.

damiano1027 committed:
refactor: 사장님 상점 메뉴 카테고리 추가/삭제 요구사항에 따른 flyway sql 파일 변경

yhsim98 commented on src/main/java/koreatech/in/service/ShopServiceImpl.java:
```diff
@@ -375,4 +375,53 @@ public Boolean shouldIncreaseShopHit(Shop shop, User user, String ip) {
+        return new HashMap<String, Object>() {{
+            put("menu_details", menuDetails);
+        }};
+    }
```
과한 주석은 가독성에 좋지 않다고 합니다
주석이 아닌 추상화를 사용하여 코드의 의도를 표현해보는 것 어떨까요?

damiano1027 replied:
어떤 코드인지를 나타내려는 과한 욕심에 주석을 많이 작성하게 되었습니다.
주석으로 설명하지 않더라도 코드에서 충분히 표현 가능하게끔 작성하는 것이 중요하다는 점 마인드셋 하고 리팩토링 진행하겠습니다.

yhsim98 commented on src/main/java/koreatech/in/service/ShopServiceImpl.java:
```diff
@@ -376,6 +373,127 @@ public Boolean shouldIncreaseShopHit(Shop shop, User user, String ip) {
+        // 이미 DB에 존재할 경우
+        if (selectedCategoryByName != null) {
+            // id가 같을때
+            if (id == selectedCategoryByName.getId()) {
```
wrapper class 간에 동등연산을 하면 어떤 문제가 생길 수 있을까요?

damiano1027 replied:
wrapper class 값은 객체이기 때문에 동일(==)연산 할 경우 메모리 주소가 비교되어 잘못 비교될 수 있습니다.
이 점을 제가 간과했네요.
equals()로 비교하는것으로 수정하겠습니다.

yhsim98 replied:
한번 직접 해보세요
수가 128 미만일 때랑, 128 이상일 때랑 한번 비교해보시면 좀 신기한 결과가 나올 겁니다.

damiano1027 replied:
128부터는 '=='의 결과가 true가 아닌 false로 나오는 것 확인했습니다.
혹시 원리가 무엇인가요?

yhsim98 replied:
wrapper 클래스는 기본값을 wrapping 할 때 valueOf 을 사용하게 됩니다. 
이때 객체 생성 비용과 메모리를 아끼기 위하여 일정 범위의 값의 인스턴스는 미리 생성해 놓고 
해당 범위 내의 기본 값이라면 미리 생성한 인스턴스를 반환하여 줍니다
그래서 동등(==)연산에서 true가 나온 거죠 같은 인스턴스니까요
이때 값의 범위는 -128~127 입니다. 참고로 범위는 static final이라 수정 못해요
범위 밖의 기본 값은 new를 통해 새로 인스턴스를 생성하여 반환해 줍니다. 
그래서 동등 연산에서 false가 나오는 겁니다.

yhsim98 replied:
그냥 null 체크 말고는 전부 equals 사용하는 것이 마음 편해요
이거 때문에 알고리즘 문제 풀다가도 여러 번 틀렸습니다. 이런 건 오류 나면 찾기도 어려워요ㅋㅋㅋ

damiano1027 replied:
아 그런 원리가 있었군요. 감사합니다 배워갑니다!

예전에 저도 C++로 알고리즘 풀때 숫자 비교를 그냥 동일 비교로 코딩하던 습관이 남아있다보니까 무의식적으로 ==를 쓰게된 것 같네요.

앞으로 저도 equals를 쓰는 습관을 가지도록 하겠습니다.

damiano1027 committed:
refactor: 사장님 상점 메뉴 Create API

- 불필요한 코드 제거
- 향상된 for문 대신 for-each 사용
- controller에서 json과 image를 동시에 받을때 json이 객체에 매핑되지 않는 문제 해결 (json의 경우 요청시 content-type을 application/json으로 해줘야함)

damiano1027 committed:
feat: 사장님 상점 메뉴 Delete API

damiano1027 committed:
feat: 사장님 상점에 존재하는 메뉴 카테고리에 대한 Read, Update API

damiano1027 committed:
refactor: 사장님 상점 메뉴 Create API

- 상점에 존재하는 메뉴 카테고리들에 대한 설정방법 변경에 따른 리팩토링

damiano1027 committed:
feat: 사장님 상점 메뉴 Update API

damiano1027 committed:
refactor: 사장님 상점 메뉴 Read API

damiano1027 committed:
refactor: 사장님 상점 메뉴 Read API 개선, 상점 메뉴 마이그레이션 API 삭제

- 메뉴 관련 테이블마다 데이터를 select하는 방식은 db 커넥션이 많이 일어나 비효율적이라 판단되었습니다.
- 그래서 관련 테이블들을 mapper xml로 1:n으로 한꺼번에 불러와 커넥션을 1번만 사용하게하고, 서비스 로직도 크게 줄였습니다.
- 상점 메뉴 마이그레이션 API를 python으로 작성하여 관리하기로하였으므로 삭제하였습니다.
- 그외 코드 네이밍을 간소화시키는 리팩토링을 진행하였습니다.

damiano1027 committed:
feat: 사장님 상점 메뉴 숨김 처리 API

damiano1027 committed:
feat: 사장님 상점 관리 페이지 메뉴 리스트 API

- 사장님 상점 관리 페이지중 메뉴 리스트를 볼 수 있는 페이지에서 필요한 데이터들을 응답해주는 API를 작성하였습니다.

damiano1027 committed:
refactor: 상점 테이블과 메뉴 카테고리 테이블간 관계 테이블을 제거함에 따라 flyway sql파일 변경

damiano1027 committed:
refactor: 상점 테이블과 메뉴 카테고리 테이블간 관계 테이블을 제거함에 따라 코드 리팩토링

damiano1027 committed:
refactor: 상점 메뉴 카테고리 추가/삭제 API, 메뉴 카테고리 등록/수정 방식 변경

damiano1027 committed:
refactor: 상점 메뉴 API

- 상점 메뉴 관련 API들을 전부 리팩토링 하였습니다. (mybatis로 리스트의 sql문 실행시 db와의 커넥션이 많이 일어나지 않도록 foreach문으로 처리)
- 상점 API는 개발중에 있습니다.

damiano1027 committed:
refactor: 상점 코드 리팩토링

- dto 패키지 추가 (domain 패키지 클래스들은 오로지 DB 테이블과의 데이터 매핑 용도로 쓰고, dto 패키지에 있는 클래스들은 요청과 응답 용도로 씀. 따라서 원하는 데이터만 요청받거나 응답 가능)
- 기존 Shop 클래스의 변경으로 인해 SearchServiceImpl.java에서 컴파일 에러가 나는 부분이 있어 주석 처리
- 기존 AdminShopController 주석 처리
- flyway sql파일에 데이터 마이그레이션 sql문 추가
- service에서 db와의 커넥션을 줄이기 위해 mybatis foreach문 도입

damiano1027 committed:
feat: 사장님 상점 매칭 API (admin 권한)

damiano1027 committed:
refactor: 상점 코드 리팩토링

- 기존 응답 타입 Map은 어떤 데이터가 응답되는지 알수 없어 가독성이 떨어짐 -> 응답 DTO 클래스를 정의해 사용하여 가독성을 높임
- 요청 DTO의 option_prices와 opens는 List<Map> 타입이라 요청 데이터의 가독성이 떨어지고 Object 객체를 각 타입으로 변환해야 하는 불편함 발생 -> OptionPrices와 Open 클래스를 정의하여 가독성을 높이고 타입 변환의 수고를 덜음
- soft delete된 상점을 어드민 권한으로도 불러올 수 없었음 -> 어드민 권한은 soft delete를 무시하고 상점 조회가 가능하게 함
- 서비스 메소드에서 자주 중복되는 코드들을 메소드로 분리

damiano1027 committed:
refactor: 요청 DTO validation 방식 수정

- 컬렉션 객체는 Controller에서 validation을 수행하지 않았던 부분을 개선
- @Min(0) 어노테이션 대신 @PositiveOrZero 사용

damiano1027 committed:
feat: 어드민 페이지 상점 리스트 조회시의 필터, 정렬, 검색

- 기존 page, limit 조건에 필터, 정렬, 검색 관련 요소를 추가
- 필터: soft delete 여부, 배달 가능 여부, 카드 계산 가능 여부, 계좌 이체 가능 여부
- 정렬: 이름순, 최신순과 오래된순
- 검색: 상점 이름

damiano1027 committed:
refactor: xssCheck() 메소드 개선

StringXssChecker.java에 있는 xssCheck()를 개선한 xssPrevent()를 동일 클래스에 추가하였습니다.

개선된 점
- 필터링 타겟 객체에 대해 Object -> Map -> Object로 변환하는 불필요한 과정을 없애고 해당 객체에서 바로 값이 필터링되어 리턴되도록 하였습니다.
- 필터링 타겟 객체의 필드 중 String 타입 필드 뿐만 아니라 List, Array, 사용자 정의 클래스(dto or domain)에 존재하는 String 객체도 필터링 될 수 있도록 하였습니다.

xssPrevent()는 controller.v2.admin.AdminShopController.java의 API만 적용해놓았습니다.

damiano1027 committed:
refactor: 어드민 페이지 상점 리스트 조회 API

- GET 요청에 대한 파라미터를 @RequestBody로 받는 것은 표준에 맞지 않아 query parameter 방식으로 변경하였습니다.

damiano1027 committed:
refactor: equals() 메소드가 재정의된 domain 클래스에 hashCode() 재정의

- `equals()`를 재정의하였을때, `hashCode()`를 재정의해주지 않으면 `Hash 기반 collection`을 사용할때 문제가 발생할 수 있으므로 재정의 하였습니다.
- `Apache Commons의 HashCodeBuilder` 라이브러리를 사용하였습니다.

damiano1027 committed:
refactor: 페이지네이션 동적 쿼리 변경

- 기존에는 동적 쿼리에서는 오직 <if> 엘리먼트만 이용하여 조건 및 필터링을 WHERE절에 추가하였지만, <trim> 엘리먼트를 이용하여 간편하게 WHERE절을 구성할 수 있게 함

damiano1027 committed:
refactor: Shop domain 클래스 리팩토링

- MyBatis SELECT시 객체 생성을 위한 NoArgsContructor 어노테이션 적용
- 메소드의 파라미터를 snake case에서 camel case로 변경

damiano1027 committed:
style: 응답 DTO 클래스들의 이름 변경

- 'Response~~~DTO' 꼴의 이름이 너무 난잡해보이고, 이미 dto 패키지에 존재하는 클래스들이기 때문에 '~~~Response' 꼴로 이름을 전부 변경

damiano1027 committed:
style: 요청 DTO 클래스들의 이름 변경

- - '~~~DTO' 꼴의 이름이 너무 난잡해보이고, 이미 dto 패키지에 존재하는 클래스들이기 때문에 '~~~Request' 꼴로 이름을 전부 변경

damiano1027 committed:
feat: 이미지 업로드 전용 API 추가

- 상점카테고리/상점메뉴/상점에 대한 이미지 업로드 전용 API를 추가
- 기존에는 상점카테고리/상점메뉴/상점에 대한 Create/Update API에서 이미지 업로드를 같이 수행하는 방식이었지만, 이를 분리시키기 위함임
- Create/Update API에서 이미지 업로드 로직 제거 필요

damiano1027 committed:
refactor: 이미지 업로드 방식 변경

- 기존 상점/상점카테고리/메뉴에 대한 생성/수정 API 호출시 이미지 업로드가 동시에 진행되는 방식에서 이미지 URL만 받는 방식으로 변경

damiano1027 committed:
feat: ExceptionMessage Enum 추가

- 기존 Exception 처리 방식은 message와 code를 직접 작성해야 하기 때문에 유지보수의 어려움 및 중복 문제 발생
- 따라서 코드를 직관적으로 작성할 수 있게 하고 중복 문제를 제거하기 위해 전용 Enum 정의

damiano1027 committed:
refactor: Exception 처리시 ExceptionMessage Enum 적용

- 기존 Exception 처리시 message와 code를 직접 입력하던 방식을 Enum으로 간편하게 처리

Choi-JJunho commented on src/main/java/koreatech/in/dto/SuccessCreateResponse.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.experimental.SuperBuilder;
+
+@Getter @SuperBuilder
+public class SuccessCreateResponse extends SuccessResponse {
```
이렇게 선언해버리면 나머지 컨트롤러들의 모든 반환부도 변경해줘야할텐데..
되려 코드의 통일성을 해치는건 아닐까요

Choi-JJunho commented on src/main/java/koreatech/in/domain/Shop/ShopMenu.java:
```diff
@@ -0,0 +1,51 @@
+    private Integer shop_id;
+    private String name;
+    private String description;
+    @Setter private Boolean is_hidden;
```
아래 reverseIsHidden() 메소드가 있는데 굳이 Setter를 붙인 이유가 있을까요?

Choi-JJunho commented on src/main/java/koreatech/in/util/ExceptionMessage.java:
```diff
@@ -0,0 +1,47 @@
+ *   - 130 ~ 139: 상점 메뉴 카테고리
+ */
+@Getter
+public enum ExceptionMessage {
```
1. 10단위가 넘어가버리면 코드가 중복될거같은데 100단위를 기준으로 하거나 알파멧을 앞에 붙여서 구분하면 어떨까요?
2. Http Status는 없나요?
3. 통일성을 주기위해 모든 코드에 대해서 에러코드를 선언해야할텐데 가능한가요?

[카카오 API 레퍼런스 참고](https://developers.kakao.com/docs/latest/ko/reference/rest-api-reference#response-code)

damiano1027 committed:
refactor: 레코드 삽입을 위한 객체 생성시 생성자가 아닌 create() static 메소드 사용

- db에 삽입될 객체임을 더 가독성 좋게 알려주기 위해 create() static 메소드에서 생성자를 호출하는 방식으로 변경하였습니다.

damiano1027 committed:
refactor: DTO 조회 방식 변경

- DTO 클래스 안에서만 사용되는 타입은 내부에서 static class로 정의하여 해당 DTO 클래스에서만 사용할 수 있게 함 (특정 화면에서 원하는 데이터만 응답하기 위해)
- ShopMapper.java의 조회 메소드 중, domain 객체를 반환하는 메소드와 DTO 객체를 반환하는 메소드가 섞여있어 혼란을 야기 --> DTO 객체 조회 전용인 ShopQueryMapper.java를 따로 만들어 해당 파일에만 DTO 객체를 조회하는 메소드를 적용
- 연관 관계가 있는 테이블 데이터를 가져올때 sub query 방식을 사용했기 때문에 N+1 문제 발생 --> JOIN으로 한번에 처리하여 쿼리 실행 횟수를 1번으로 줄임

damiano1027 committed:
refactor: LinkedList를 전부 ArrayList로 변경

damiano1027 committed:
refactor: PUT API의 정상 응답을 201에서 200으로 교체

damiano1027 committed:
refactor: 수정 API에서의 이미지 url 및 카테고리에 관한 로직 수정

- DB에서 DELETE할 객체들은 (기존 리스트)에서 (요청된 리스트)를 제외했을때 남는 객체들로 판단
- DB에 INSERT할 객체들은 (요청 리스트)에서 (기존 리스트)를 제외했을때 남는 객체들로 판단
- Set으로 처리해보고자 하였으나 차집합(removeAll) 처리가 List의 removeAll과 비슷하게 동작하기 때문에 기존 방식인 List로 처리

damiano1027 committed:
fix: INSERT된 레코드의 id를 가져오지 못하는 문제 수정

- property를 param 어노테이션에서 설정한 이름에 '.'를 붙이고 필드명을 붙여 해결

damiano1027 committed:
refactor: 메뉴 옵션 중복 로직

- 메뉴의 어떤 옵션명이 중복됐는지 확인하는 것이 아닌, 단순히 중복 여부를 확인하므로 set을 이용해서 set에 담긴 개수와 요청 개수가 다르면 중복으로 처리

damiano1027 committed:
refactor: 정상 응답 방식 변경

- 기존에는 SuccessResponse와 SuccessCreateResponse를 이용해서 200번대 응답시 body에 데이터를 응답하였지만, 정책 변경으로 body를 없앤 방식으로 변경
- 정상 응답과 관련된 API의 Service 메소드는 리턴 타입을 void로 하고, Controller 메소드의 리턴 타입은 ResponseEntity<Void>로 하여 body 없이 http 상태 코드만 응답

damiano1027 committed:
refactor: MapStruct 적용

DTO <-> 엔티티(DB에서 데이터를 가져와 매핑한 객체)간 손쉬운 데이터 매핑을 위해 MapStruct 도입

damiano1027 committed:
chore: 상점 리뉴얼 API를 V2에서 기본으로 변경

damiano1027 committed:
Merge branch 'develop' into feature/shop-menu-renewal

# Conflicts:
#	pom.xml
#	src/main/java/koreatech/in/controller/ShopController.java
#	src/main/java/koreatech/in/controller/admin/AdminShopController.java
#	src/main/java/koreatech/in/domain/Criteria/Criteria.java
#	src/main/java/koreatech/in/service/ShopServiceImpl.java
#	src/main/java/koreatech/in/util/StringXssChecker.java
#	src/main/resources/common/context-mapper.xml

damiano1027 committed:
feat: 특정 테이블 unique index 적용하는 flyway 파일 추가

damiano1027 committed:
refactor: 패키지 구조 변경

- dto, mapstruct 패키지에서 하위 패키지에 admin/normal 패키지를 두고 그 하위에 각 도메인에 대한 패키지를 두는 방식으로 변경
- controller, service, repository 패키지 구조와 통일시키기 위함

damiano1027 committed:
docs: 상점 카테고리 API Swagger 문서 정리

- swagger 문서 정리
- 자잘한 에러 수정

damiano1027 committed:
docs: 상점 메뉴 카테고리 API Swagger 문서 정리

damiano1027 committed:
docs: 상점 카테고리 API Swagger 문서 수정

damiano1027 committed:
docs: 상점 API Swagger 문서 정리

damiano1027 committed:
docs: 상점 메뉴 API Swagger 문서 정리

damiano1027 committed:
docs: Swagger에 401, 403 응답에 대한 설명 추가

damiano1027 committed:
Merge pull request #131 from BCSDLab/task/admin-shop-api-swagger

Task: 어드민 상점 API Swagger 문서 정리

damiano1027 committed:
feat: 사장님 상점 API 전용 Controller, Service 등록

damiano1027 committed:
feat: 메뉴 카테고리 생성 dto 정의

damiano1027 committed:
feat: 메뉴 카테고리 생성 API

damiano1027 committed:
feat: 특정 상점의 모든 메뉴 카테고리 조회 API

damiano1027 committed:
feat: 메뉴 카테고리 삭제 API

damiano1027 committed:
docs: Swagger 설명 추가

damiano1027 committed:
Merge pull request #132 from BCSDLab/task/owner-shop-menu-category-api

Task: 사장님 전용 메뉴 카테고리 API

damiano1027 committed:
refactor: root uri를 controller 클래스 상단에 선언

damiano1027 committed:
docs: 모든 상점 카테고리 조회 API Swagger 문서 정리

damiano1027 committed:
docs: 특정 상점 조회 API Swagger 문서 정리

damiano1027 committed:
docs: 모든 상점 조회 API Swagger 문서 정리

damiano1027 committed:
fix: 상점의 모든 메뉴 조회시 상점 존재 여부를 확인하지 않던 문제 수정

damiano1027 committed:
docs: 특정 상점의 모든 메뉴 조회 API Swagger 문서 정리

damiano1027 committed:
docs: 특정 상점 조회 API Swagger 설명에 error code 누락부분 추가

damiano1027 committed:
feat: 상점 카테고리 이미지 url에 https:// 추가

damiano1027 committed:
fix: sql 파일에서 mysql5에서 지원하지 않는 문법을 사용한 문제 수정

damiano1027 committed:
feat: 전체보기 상점 카테고리에 대한 관계 추가 sql 파일 작성

damiano1027 committed:
Merge pull request #133 from BCSDLab/task/shop-api-swagger

Task: 일반 상점 API Swagger 문서 정리 및 Flyway 파일 추가

damiano1027 committed:
fix: Swagger에서 path variable 값이 요청에 들어가지 않던 문제 수정

- PathVariable 파라미터에 ApiParam 선언과 name 속성을 넣어주게되면 Swagger에서 path variable 값이 요청에 들어가지 않았음.
- 따라서 ApiParam 어노테이션에서 name 속성은 사용하지 않는 것으로 수정

damiano1027 committed:
fix: Swagger에서 422 응답을 412 응답으로 잘못 표기했던 문제 수정

damiano1027 committed:
Merge pull request #134 from BCSDLab/task/admin-shop-api-swagger-2

Task: 어드민 상점 API Swagger 수정

damiano1027 committed:
Merge branch 'develop' into feature/shop-menu-renewal

damiano1027 committed:
feat: 메뉴 생성 API

damiano1027 committed:
feat: 메뉴 조회 API

damiano1027 committed:
feat: 상점의 모든 메뉴 조회 API

damiano1027 committed:
refactor: AllMenusOfShopResponse dto 생성시에는 size와 리스트만 변환하는 역할만 하게 수정

damiano1027 committed:
refactor: Swagger에서 dto의 불필요한 getter를 인식하지 못하게 수정

damiano1027 committed:
feat: 메뉴 삭제 API

damiano1027 committed:
feat: 메뉴 수정 API

damiano1027 committed:
Merge pull request #135 from BCSDLab/task/owner-shop-menu-api

Task: 사장님 전용 상점 메뉴 API

damiano1027 committed:
Merge branch 'develop' into feature/shop-menu-renewal

damiano1027 committed:
docs: Swagger에서의 422 응답 스키마 변경

damiano1027 committed:
refactor: 상점 및 상점 카테고리 API에 대한 422 응답 메시지 변경

damiano1027 committed:
refactor: 메뉴 및 메뉴 카테고리에 대한 422 응답 메시지 변경

damiano1027 committed:
docs: 사장님 전용 상점 API Swagger에서의 422 응답 스키마 변경

damiano1027 committed:
refactor: 사장님 전용 상점 API 422 응답에 대한 메시지 변경

damiano1027 committed:
fix: 유저 sql과 merge후 충돌 방지를 위한 flyway sql 버전 수정

damiano1027 committed:
Merge pull request #140 from BCSDLab/task/refactor-shop-api-for-changed-422-response

Task: 상점 API의 422 응답에 대한 제약조건 위반 메시지 수정

damiano1027 committed:
Merge branch 'develop' into feature/shop-menu-renewal

# Conflicts:
#	src/main/java/koreatech/in/util/StringXssChecker.java
